### ISSUE TITLE
Support keyword index-time normalizers and allow multiplexer

### ISSUE THREAD

**Issue by mitar** [elastic/elasticsearch#83004]
Currently, keyword fields can define normalizers. They change the keyword before it is indexed and they also change input terms during search time. I would propose that normalizers are split into index-time normalizers and search-time normalizers (similar to how analyzers can be done). This would allow one to use `multiplexer` filter during index-time which could produce multiple keywords (which is great, they should all be indexed) and then during search time provided input would be matched against any of those produced keywords.

My use case is that I have a field which is storing IDs. Those IDs might be coming from different sources, but in general there might be noise. There might be an ID `Abc123`, but one cannot really know if search for `abc123` should match it or not. Maybe this particular ID format is case sensitive, or maybe it is not. So ideally I would use `multiplexer` normalizer which would use `lowercase`, or maybe even `decimal_digit,asciifolding,lowercase` filters. Normalizers might then produce multiple values which would be indexed. If user then searches for `abc123` I could then find `Abc123` as an option, if there is nothing better. But if user searches for `Abc123`, I would find only `Abc123`, even if there is `abc123` as well (remember, normalization is done only during index time, there is no case changing during search time).

**Comment by javanna** [comment]
hi @mitar , normalizers are allowed to emit a single token, and that is why they can be applied to keyword fields in contrast to analyzers. I am under the impression that with the change you are proposing, normalizers would become the same as analyzers though. I am probably missing something, could you please clarify this?

**Comment by mitar** [comment]
> I am under the impression that with the change you are proposing, normalizers would become the same as analyzers though. I am probably missing something, could you please clarify this?

Yes, with my proposal the index-time normalizers could return multiple tokens then, yes. So yea, maybe the easiest would be to just make them analyzers. So my proposal would then be:

* Make index-time normalizers be analyzers.
* Search-time normalizers stay normalizers.
* If you define just one, then it has to be a normalizer, which is used for both index and search time (current state).

**Comment by javanna** [comment]
heya @mitar thanks for clarifying your feature request. The reason why normalizers can only emit one token is that their values can then be loaded into doc_values to speed up aggregations/sorting and in general accessing their values by doc id. That is why normalizers differ from analyzers, as well as the reason why text fields don't support doc_values and require fielddata to perform aggregations on them, which is very expensive to load, while keyword fields have doc_values enabled by default. I am going to close this as won't fix, as we are not going to make normalizers return multiple tokens.

**Comment by mitar** [comment]
Hm, but keyword fields can already be arrays, by default? So what I am asking here is to have an implicit index-time array generation (if normalizer returns more than one token). Everything else stays the same. So instead of me indexing:

```
keywordField: ["ABC", "abc"]
```

I could just index:

```
keywordField: "ABC"
```

And then `multiplexer` would apply `lowercase` filter to produce `["ABC", "abc"]`. So technically there is not much difference. The only difference is that this is done at index time with `_source` not having to be changed itself. Similarly how one does not have to change `_source` to tokenize/stem all words in text search, but can use an analyzer.

**Comment by javanna** [comment]
I see what you mean. We've historically rejected this type of requests, and decided to keep a clear distinction between keyword and text fields. I am not against re-evaluating this one more time with the team though, just to see if things have changed that would make us reconsider this decision. Marking for discussion.

**Comment by mitar** [comment]
Thanks!

**Comment by benwtrent** [comment]
The guarantee that a normalizer produces a single token from a single token is utilized within the codebase extensively. 

We wouldn't want to change this. The additional complexity, not just cognitively in the API (now, what would be the point of normalizers at all...), but within the codebase is difficult to justify when this request seems achievable in various other ways.


I would say the way to achieve "indexing a keyword in multiple ways" is with "multi-fields" and then searching those multiple fields at a time. 


Another option for more complex analysis is a `text` field and then using `term` filtering on the text field.
